sympy/calculus/util.py

```diff
@@ -10,6 +10,131 @@
+    (1/2, oo)
+
+    """
+    from sympy.solvers.inequalities import solve_univariate_inequality
+    from sympy.solvers.solveset import solveset, _has_rational_power
+
+    if domain.is_subset(S.Reals):
+        constrained_interval = domain
+        for atom in f.atoms(Pow):
+            predicate, denom = _has_rational_power(atom, symbol)
+            constraint = S.EmptySet
+            if predicate and denom == 2:
+                constraint = solve_univariate_inequality(atom.base >= 0,
+                                                         symbol).as_set()
+                constrained_interval = Intersection(constraint,
+                                                    constrained_interval)
+        for atom in f.atoms(log):
+            constraint = solve_univariate_inequality(atom.args[0] > 0,
+                                                     symbol).as_set()
+            constrained_interval = Intersection(constraint,
+                                                constrained_interval)
+        domain = constrained_interval
+    try:
+        sings = S.EmptySet
+        for atom in f.atoms(Pow):
+            predicate, denom = _has_rational_power(atom, symbol)
+            if predicate and denom == 2:
+                sings = solveset(1/f, symbol, domain)
+                break
+        else:
+            sings = Intersection(solveset(1/f, symbol), domain)
+    except:
+        raise NotImplementedError("Methods for determining the continuous domains"
+                                  " of this function has not been developed.")
+
+    return domain - sings
+
+
+def function_range(f, symbol, domain):
+    """
+    Finds the range of a function in a given domain.
+    This method is limited by the ability to determine the singularities and
+    determine limits.
+
+    Examples
+    ========
+
+    >>> from sympy import Symbol, S, exp, log, pi, sqrt, sin, tan
+    >>> from sympy.sets import Interval
+    >>> from sympy.calculus.util import function_range
+    >>> x = Symbol('x')
+    >>> function_range(sin(x), x, Interval(0, 2*pi))
+    [-1, 1]
+    >>> function_range(tan(x), x, Interval(-pi/2, pi/2))
+    (-oo, oo)
+    >>> function_range(1/x, x, S.Reals)
+    (-oo, oo)
+    >>> function_range(exp(x), x, S.Reals)
+    (0, oo)
+    >>> function_range(log(x), x, S.Reals)
+    (-oo, oo)
+    >>> function_range(sqrt(x), x , Interval(-5, 9))
+    [0, 3]
+
+    """
+    from sympy.solvers.solveset import solveset
+
+    vals = S.EmptySet
+    intervals = continuous_domain(f, symbol, domain)
+    range_int = S.EmptySet
+    if isinstance(intervals, Interval):
+        interval_iter = (intervals,)
+    else:
+        interval_iter = intervals.args
+
+    for interval in interval_iter:
+        cps = S.EmptySet
+        cvs = S.EmptySet
+        bounds = ((interval.left_open, interval.inf, '+'),
+                  (interval.right_open, interval.sup, '-'))
+
+        for i in bounds:
```

Comment:
In general accessing element of a heterogeneous tuple by index can be very confusing. In such cases also consider using `namedtuples` https://www.youtube.com/watch?v=GfxJYp9_nJA